### ISSUE TITLE
will execute the same yaml benchmark on different GPUS ON THE same machine overwrite existing CSVS? 

### ISSUE THREAD

**Issue by smilesun** [marrlab/DomainLab#186]
If we run `./run_benchmark.sh ./examples/yaml/demo_benchmark.yaml 0 ` in the DomainLab repo, then we open another command line session, in the same folder, and run `./run_benchmark.sh ./examples/yaml/demo_benchmark.yaml 1`, can we ensure the results for both run will be saved in csv files? 

For nonbenchmark run, results are just appended into the csv file, not a problem how about for benchmark?

The reason is we have limited computational resources, by running `./run_benchmark.sh ./examples/yaml/demo_benchmark.yaml SEED ` , we want to sample different hyperparameters so that we can plot them together with  `python main_out.py --agg_partial_bm OUTPUT_DIR`

**Comment by Car-la-F** [comment]
@smilesun,

Both sessions will try to write inside the same .csv files. As these files are appended with every result this should in general not overwrite the results of the other session.

https://github.com/marrlab/DomainLab/blob/860cb4a28e993124dd8bff63a175078e8902abdf/domainlab/compos/exp/exp_utils.py#L182

However, if two processes try to write inside the same file at the same time the line will get messed up. I also assume we will get the header twice in each file.

---

A solution for this might be to include the random seed in the name of the .csv files, so each session will get their own .csv files inside the same _rule_results_ folder. From the implementation of the `agg_partial_bm` I know that it does not matter how many files are created inside the _rule_results_ folder, so I think this would be the most convenient solution to this problem.

**Comment by Car-la-F** [comment]
@smilesun Shall I try to implement this solution?

**Comment by smilesun** [comment]
1. sounds good, although the scenario you mentioned has low probability to occur (maybe hight probability) i think it is good to avoid it if the simply solution by adding seed to csv file name could help. 

2.  for me an extra scenario in mind is if we exeucte `./run_benchmark.sh ./examples/yaml/demo_benchmark.yaml 0 ` after it finishes, someone execute `./run_benchmark.sh ./examples/yaml/demo_benchmark.yaml 1`, what will happen in this case?   I assume the first finished benchmark will delete already the  individual csv files?

**Comment by Car-la-F** [comment]
@smilesun 

1. Okay, then I will work on this

2. I looked back at the benchmarks we did for the project. It seems like the partial results are not deleted after they are aggregated into one file. I also looked inside the code and couln't find any code which seems to delete any of these files. 

From this I came to the conclusion that if one executes benchmarks like you descrieb in 2.:
- `./run_benchmark.sh ./examples/yaml/demo_benchmark.yaml 0` will generate partial results in the _rule_results_ folder and aggregate them into a `results.csv` file after the benchmark is finished. The partial results are not deleted after the aggregation
- the second execution `./run_benchmark.sh ./examples/yaml/demo_benchmark.yaml 1` will then append the existing partial results .csv files in the _rule_results_ folder. Hence they will contain results from both benchmarks. The aggregation in the end is implemented such that it overwrites existing .csv files. Hence the resulting aggregated result will contain all results from both benchmarks.

---

Regarding the issue with the two headers. It seems like this will not happen, as the header is only created if the file does not exist yet. A header is only created if the following line gives `True`:

https://github.com/marrlab/DomainLab/blob/860cb4a28e993124dd8bff63a175078e8902abdf/domainlab/compos/exp/exp_utils.py#L123

**Comment by Car-la-F** [comment]
Seems like the issue with running two sessions in snakemake is, that skemake does lock all files and directories used by the session. Hence they cannot be accessed by another session at the same time 

```
Error: Directory cannot be locked. Please make sure that no other Snakemake process is trying to create the same files in the following directory:
/home/carla/Dokumente/Uni/Master/Computational_Methods_for_Single-Cell_Biology/DomainLab
If you are sure that no other instances of snakemake are running on this directory, the remaining lock was likely caused by a kill signal or a power loss. It can be removed with the --unlock argument.
```

I still get this issues after renaming the files for the partial results, the hyperparameters, the full results and the graphics (by using the random seed). 

---

@smilesun 
Before spending more time in solving this issue I was wondering about the advantage of starting two sessions with one hyperparameter sample per algorithm, over starting just one session with two hyperparameter samples per algorithm. 

You wrote:
"The reason is we have limited computational resources, by running `./run_benchmark.sh ./examples/yaml/demo_benchmark.yaml SEED`"
do you know where this does come from? Maybe it would be easier to solve this issue than getting snakemake to allow two processes working on the same files.

**Comment by smilesun** [comment]
If this will take a lot of time to resolve, then we should patiently wait until one machine to finish all jobs/experiments.

The issue is if we want to run a large benchmark without a slurm cluster but only one machine with several GPUs, to have result within 1 week instead of 1 month, we might want to unitize all 8 GPUs to let them run different jobs/experiments of the benchmark instead of only letting one GPU to do all the work. 

That being said, it sounds like we have chosen the wrong issue? A solution that makes more sense is to investigate how to let snakemake to find all available GPUs and dispatch different jobs of the benchmark into different GPUs so all 8 GPUs on the same machine will be busy?

**Comment by Car-la-F** [comment]
@smilesun 

From what I know, it is possible to specify the number of cores in the run_benchmark.sh file 

https://github.com/marrlab/DomainLab/blob/860cb4a28e993124dd8bff63a175078e8902abdf/run_benchmark.sh#L28

right now this is hard coded to be one (`--cores 1`), but one can adapt this depending on the machine which is used.
Maybe it does make sense to add this as a command line argument.

The only thing to be careful with this is, that even if a machine has a specific number of cores, the process may crash when using all cores if the GPU memory isn't big enough. But this might be the same problem when starting multiple sessions at the same time.

**Comment by smilesun** [comment]
I think the "cores" here refer to the CPU cores, not how many GPUs to use

**Comment by Car-la-F** [comment]
@smilesun 

A second thing I realized when running `./run_benchmark.sh ./examples/yaml/demo_benchmark.yaml 0 `and `./run_benchmark.sh ./examples/yaml/demo_benchmark.yaml 1`:
It seems like both sample the same hyper parameters, because they use the same random seed, which does again explain, why the snakemake sessions do access the same files (after adding the random seed in the file name). 

I did find that that the number in the end of the command line arguments do set the `DOMAINLAB_CUDA_START_SEED`

https://github.com/marrlab/DomainLab/blob/860cb4a28e993124dd8bff63a175078e8902abdf/run_benchmark.sh#L16

This variable is, however, not passed to the snakemake session. In the snakemake session the random seed specified in the benchmark yaml is used.
What is the goal of using the `export DOMAINLAB_CUDA_START_SEED=$2`?

**Comment by smilesun** [comment]
because we have
1. https://github.com/marrlab/DomainLab/blob/860cb4a28e993124dd8bff63a175078e8902abdf/domainlab/exp_protocol/benchmark.smk#L19
2. https://github.com/marrlab/DomainLab/blob/860cb4a28e993124dd8bff63a175078e8902abdf/domainlab/exp_protocol/benchmark.smk#L65
then we could use different random seed on different machines

**Comment by Car-la-F** [comment]
> because we have
> 
>     1. https://github.com/marrlab/DomainLab/blob/860cb4a28e993124dd8bff63a175078e8902abdf/domainlab/exp_protocol/benchmark.smk#L19
> 
>     2. https://github.com/marrlab/DomainLab/blob/860cb4a28e993124dd8bff63a175078e8902abdf/domainlab/exp_protocol/benchmark.smk#L65
>        
>        then we could use different random seed on different machines

I see, I think I misunderstood the usage of `DOMAINLAB_CUDA_START_SEED`, it does control the stochastic variation and not the systematic variation.

**Comment by smilesun** [comment]
Ah, OK, I guess then we need to set a different random seed with corresponding environment variable to sample different hyper-parameters?  If we sample different hyper-parameters, then it will be possible to run two different sessions in side the same folder?

**Comment by Car-la-F** [comment]
> Ah, OK, I guess then we need to set a different random seed with corresponding environment variable to sample different hyper-parameters? If we sample different hyper-parameters, then it will be possible to run two different sessions in side the same folder?

I think we are still phasing the same issue then. Both session will need to work on completly different files. Hence we cannot have one results.csv file after the sessions finish.

I would rather investigate wheather it is possible to use multiple GPUs with one session. 
Looking into the snakemake documentation it seam prommising to specify the number of GPUs in the rule. e.g.:

```
input:
        # path to config file as input, thus a full
        # rerun is considered whenever the config yaml changed.
        expand("{path}", path=config_path)
    output:
        dest=expand("{output_dir}/hyperparameters_" + str(os.environ["DOMAINLAB_CUDA_START_SEED"]) + ".csv",
            output_dir=config["output_dir"])
    resources:
        gpu=1
    run:
        from domainlab.utils.hyperparameter_sampling import sample_hyperparameters
        sample_hyperparameters(config, str(output.dest))
```

https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#gpu-resources

**Comment by smilesun** [comment]
then i guess we can create a branch with 
`nvidia_gpu=3`  added in the smk file and let our collaborator who has a 8 GPU machine to test if this works?

**Comment by Car-la-F** [comment]
Sounds good.

I was thinking about adding the number of GPUs as a comand line argument or is it better to put it directly inside the smk file?

**Comment by smilesun** [comment]
if we can do it more general , sure, it will be better and we can even merge it to master, and then let our collaborator test it.

**Comment by Car-la-F** [comment]
I added the resources to all jobs. I tested the code (with the limited resources of my Laptop), so it does not crash. 

The code can be found on the branch `cf_multiple_GPU`. One can execute a benchmark yaml with 

```
./run_benchmark.sh <CONFIGFILE> <DOMAINLAB_CUDA_START_SEED> <NUMBER_GPUS>
```

I did use `CONFIGFILE = ./examples/yaml/demo_benchmark.yaml` as this is a very small benchmark. But maybe, when using multiple GPUs it is better to modify the yaml to use more epos and more hyper parameter samples.

**Comment by Car-la-F** [comment]
By the way, do we still whant to use the DOMAINLAB_CUDA_START_SEED? It can be confusing because it is also set in the benchmark yaml file

**Comment by smilesun** [comment]
Do you mean we have start_seed and end_seed already in the YAML file ?
If DOMAINLAB_CUDA_SEED is not empty, the start_seed will be overwritten and
end seed will be start seed plus the difference in the YAML file.

It is still designed in this way because we want to let different machines
to run the same benchmark with different random seed.

Does that make sense?

**Comment by Car-la-F** [comment]
Okay, that makes sense to me

**Comment by smilesun** [comment]
in this commit, i have tested different seed lead to different hyper-parameters, but if i run the same benchmark with two different hyper-parameters, the old results get over-written by the new benchmark
[90e44b2](https://github.com/marrlab/DomainLab/commit/90e44b2cc7e85522630141874079fa042804cadc)

```
./run_benchmark.sh examples/yaml/demo_benchmark.yaml 10 101
./run_benchmark.sh examples/yaml/demo_benchmark.yaml 10 102
```
the seed with 102 overwrite the one with 101

**Comment by Car-la-F** [comment]
Yes, because they are saved in the same output directory. I think the whole pipeline is not meant to run the same .yaml file multiple times on the same machine.

The idea of allowing multiple GPU work on the same benchmark is also to not be forced run a benchmark multiple times with different seeds. 

Maybe (to prevent future confusion) we could add a time stamp in the end of the name of the output file. So none of the already existing results can be overwritten?

> in this commit, i have tested different seed lead to different hyper-parameters, but if i run the same benchmark with two different hyper-parameters, the old results get over-written by the new benchmark [90e44b2](https://github.com/marrlab/DomainLab/commit/90e44b2cc7e85522630141874079fa042804cadc)
> 
> ```
> ./run_benchmark.sh examples/yaml/demo_benchmark.yaml 10 101
> ./run_benchmark.sh examples/yaml/demo_benchmark.yaml 10 102
> ```
> 
> the seed with 102 overwrite the one with 101
